akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterPortsSpec.scala

```diff
@@ -12,1166 +12,1529 @@ class GraphInterpreterPortsSpec extends StreamSpec with GraphInterpreterSpecKit
+        out.isAvailable should be(false)
+        out.isClosed should be(true)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          out.push(0)
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          in.grab()
+        }
+
+        out.complete() // This should have no effect now
+        stepAll()
+
+        lastEvents() should be(Set.empty)
+        out.isAvailable should be(false)
+        out.isClosed should be(true)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          out.push(0)
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          in.grab()
+        }
+
+        in.cancel() // This should have no effect now
+        stepAll()
+
+        lastEvents() should be(Set.empty)
+        out.isAvailable should be(false)
+        out.isClosed should be(true)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          out.push(0)
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          in.grab()
+        }
+      }
+
+      s"propagate cancel while push is in flight (chasing = $chasing)" in new PortTestSetup(chasing) {
+        in.pull()
+        stepAll()
+        clearEvents()
+
+        out.push(0)
+
+        lastEvents() should be(Set.empty)
+        out.isAvailable should be(false)
+        out.isClosed should be(false)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(true)
+        in.isClosed should be(false)
+
+        in.cancel()
+
+        lastEvents() should be(Set.empty)
+        out.isAvailable should be(false)
+        out.isClosed should be(false)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+
+        stepAll()
+
+        lastEvents() should be(Set(Cancel(out)))
+        out.isAvailable should be(false)
+        out.isClosed should be(true)
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          out.push(0)
+        }
+        an[Ille
```

Comment:
Yeah, once I got burnt with these corner cases I created this suite. Unfortunately it didn't end up triggering the chasing behavior. The current modifications also don't cover the things I found, I had to do more specialized test in ChasingEventsSpec, but extra coverage never hurts.